torch/_higher_order_ops/utils.py

```diff
@@ -546,3 +538,62 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
     assert all(
         isinstance(arg, (torch.Tensor, int, torch.SymInt)) for arg in lifted_args
     ), f"{lifted_args} can only be of {allowed_types} but got {tuple(type(arg) for arg in lifted_args)}"
+
+
+def check_input_alias_and_mutation(
+    gm: torch.fx.GraphModule,
+    fake_args: List[FakeTensor],
+) -> Tuple[List[int], dict[int, int], dict[int, int]]:
+    with disable_proxy_modes_tracing():
+        """This function creturns mutated inputs, inp-out alias and out-out alias
+        in the graph module gm.
+        """
+
+        def _tensor_version(t) -> Optional[int]:
+            if isinstance(t, FakeTensor):
+                return t._version
+            return None
+
+        def _tensor_storage(t) -> StorageWeakRef:
+            return StorageWeakRef(t._typed_storage())
+
+        # Clone the fake args to avoid mutating the original fake args
+        cloned = [
+            arg.clone() if isinstance(arg, torch.Tensor) else arg for arg in fake_args
+        ]
+        before = [_tensor_version(arg) for arg in cloned]
+        # We need to temporarily turn inference_mode off because
+        # under inference mode, tensor version counter is not tracked.
+        fake_mode = torch._guards.detect_fake_mode(fake_args)
+        fake_ctx = fake_mode if fake_mode is not None else nullcontext()
+        ignore_fresh_unbacked = (
+            fake_mode.shape_env.ignore_fresh_unbacked_symbols()
+            if fake_mode is not None
+            else nullcontext()
+        )
+        with torch.inference_mode(False), fake_ctx, ignore_fresh_unbacked:
+            outputs = gm(*cloned)
```

Comment:
Do you have a test for this?

Also, I think the cloned tensors need to be closed when inference_mode is off. Otherwise, they may not be eligible to have version counters (tensors created under inference mode do not have version counters)

Reply:
yeah, added tests.